### PR TITLE
Allow `Listenable.merge()` to use any iterable

### LINKED ISSUE #143664: _MergingListenable should use an Iterable
### Use case

The `_MergingListenable` in [change_notifier.dart](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/foundation/change_notifier.dart) uses a `List` to store its children. If the constructor is called with an `Iterable` for the children, it must be converted to a `List` unnecessarily.

### Proposal

The type of `_MergingListenable._children` should be changed to `Iterable<Listenable?>`, as it is only used for iteration through for-loops.

### PR DESCRIPTION
**Pull request by nate-thegrate**
This is a very small change that fixes #143664.

### CODE REVIEW

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 1]
```diff
@@ -62,11 +62,11 @@ abstract class Listenable {
   /// Return a [Listenable] that triggers when any of the given [Listenable]s
   /// themselves trigger.
   ///
-  /// The list must not be changed after this method has been called. Doing so
-  /// will lead to memory leaks or exceptions.
+  /// [listenables] must not be changed after this method has been called.
```
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/change_notifier.dart:65]: nit: Per style guide, docs should start grammatically correct with a capital letter, see https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use-correct-grammar.

May have to reword this sentence slightly to make that possible here.

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 2]
```diff
@@ -62,11 +62,11 @@ abstract class Listenable {
...
+  /// Doing so will lead to memory leaks or exceptions.
   ///
-  /// The list may contain nulls; they are ignored.
-  factory Listenable.merge(List<Listenable?> listenables) = _MergingListenable;
+  /// The iterable may contain nulls; they are ignored.
+  factory Listenable.merge(Iterable<Listenable?> listenables) = _MergingListenable;
```
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/change_notifier.dart:69]: Can you add a simple test for this change (e.g. a test that passes in a statically typed Iterable and would break if we ever revert this change)?

File: packages/flutter/lib/src/foundation/change_notifier.dart (hunk 1 above)
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/change_notifier.dart:65]: Actually, this documentation could be misunderstood that you are not allowed to change the individual `Listenables`. We should make it clear that you are not allowed to change what the Iterable iterates over. Every iteration must yield the same result.

File: packages/flutter/lib/src/foundation/change_notifier.dart (hunk 2 above)
Comment (nate-thegrate) [reply packages/flutter/lib/src/foundation/change_notifier.dart:69]: Sounds good! There's already a test that uses multiple `List`s, so I changed one of them to a `Set`.

File: packages/flutter/lib/src/foundation/change_notifier.dart (hunk 1 above)
Comment (nate-thegrate) [reply packages/flutter/lib/src/foundation/change_notifier.dart]: Yeah, that's a good point. Hopefully the current documentation is better:

```dart
/// Once the factory is called, items must not be added or removed from the iterable.
/// Doing so will lead to memory leaks or exceptions.
```

File: packages/flutter/test/foundation/change_notifier_test.dart [hunk 3]
```diff
@@ -282,7 +282,7 @@ void main() {
     final TestNotifier source3 = TestNotifier();
     final List<String> log = <String>[];
 
-    final Listenable merged = Listenable.merge(<Listenable>[source1, source2]);
+    final Listenable merged = Listenable.merge(<Listenable>{source1, source2});
```
Comment (goderbauer) [inline packages/flutter/test/foundation/change_notifier_test.dart:285]: My preference would be to add a new test for this specific case so it is clear what we want to test. The test could be named "ensure that merging listenables works with non-lists" or something like that.

**Review by goderbauer — APPROVED** [review]
LGTM otherwise

**Review by goderbauer — APPROVED** [review]
LGTM

**Review by cbracken — APPROVED** [review]

**Review by goderbauer — APPROVED** [review]
LGTM

**Comment by nate-thegrate** [comment]
Thank you!
